empty-bin.lua

```diff
@@ -3,24 +3,51 @@
 -- https://gist.github.com/stonetoad/11129025
 -- http://dwarffortresswiki.org/index.php/DF2014_Talk:Bin
 
---[====[
-
-empty-bin
-=========
+local function moveItem(item, to_pos)
+    print('  ' .. dfhack.items.getDescription(item, 0))
+    dfhack.items.moveToGround(item, to_pos)
+end
 
-Empties the contents of the selected bin onto the floor.
+local function emptyContainer(container)
+    local items = dfhack.items.getContainedItems(container)
 
-]====]
+    if #items > 0 then
+        print('Emptying ' .. dfhack.items.getDescription(container, 0))
+        for _, item in ipairs(items) do
+            moveItem(item, xyz2pos(dfhack.items.getPosition(container)))
+        end
+    end
+end
 
-local bin = dfhack.gui.getSelectedItem(true) or qerror("No item selected")
-local items = dfhack.items.getContainedItems(bin)
+local viewsheets = df.global.game.main_interface.view_sheets
 
-if #items > 0 then
-    print('Emptying ' .. dfhack.items.getDescription(bin, 0))
-    for _, item in pairs(items) do
-        print('  ' .. dfhack.items.getDescription(item, 0))
-        dfhack.items.moveToGround(item, xyz2pos(dfhack.items.getPosition(bin)))
+local stockpile = dfhack.gui.getSelectedStockpile(true)
+local selectedItem = dfhack.gui.getSelectedItem(true)
+local selectedBuilding = dfhack.gui.getSelectedBuilding(true)
+if stockpile then
+    local contents = dfhack.buildings.getStockpileContents(stockpile)
+    for _, container in ipairs(contents) do
+        emptyContainer(container)
+    end
+elseif selectedItem then
+    moveItem(selectedItem, xyz2pos(dfhack.items.getPosition(selectedItem)))
+    emptyContainer(selectedItem)
+elseif selectedBuilding then
+    if not df.building_actual:is_instance(selectedBuilding) then
+        qerror("This type of building does not contain any items!")
+    end
+    local containedItems = selectedBuilding.contained_items
+    for i=#containedItems-1,0,-1 do
+        contained = containedItems[i]
+        if contained.use_mode == df.building_item_role_type.TEMP then
+            moveItem(contained.item, xyz2pos(selectedBuilding.centerx, selectedBuilding.centery, selectedBuilding.z))
```

Comment:
why are we moving the items out of the building instead of dumping the contents of contained containers?